arch/x86/core/intel64/locore.S

```diff
@@ -78,15 +78,41 @@
 	movq %rax, %cr0
 .endm
 
+/* The .locore section begins the page-aligned initialization region
+ * of low memory.  The first address is used as the architectural
+ * entry point for auxiliary CPUs being brought up (in real mode!)
+ * via a startup IPI.  It's is ALSO used by some loaders (well,
+ * ACRN...) who hard-coded the address by inspecting _start on a
+ * non-SMP build.
+ *
+ *               === OUTRAGEOUS HACK FOLLOWS ===
+ *
+ * Therefore it needs to start at OS entry with a 32 bit jump to the
+ * 32 bit entry point, and gets clobbered later (see the beginning of
+ * __start32) with NOP bytes such that the next CPU will fall through
+ * to the 16 bit SMP entry.
+ *
+ * We write out the JMP followed by 8 NOPs for simplicity.  No i386
+ * JMP encodes with more than 8 bytes, so we can come back later and
+ * scribble over it with 8 0x90 bytes (which is the 1-byte NOP) and be
+ * sure to get all of it without overwriting anything.
+ */
 .section .locore,"ax"
+.code32
+.globl __start
+__start:
+	jmp __start32
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
```

Comment:
Can't you just check if the proc is in protected mode using `smsw` and jump to `__start32`, using instruction encodings that are common to both real and protected modes? (after all, it doesn't matter whether we use `eax` or `ax`, since `PE` is the LSB).

Reply:
> The MSW register has other problems
or `mov eax, cr0` for that matter; it's `0F20C0` in both real and protected modes.

Reply:
Touching EAX in real mode requires an address size prefix, no?  I really don't see that kind of trickery as an improvement.  The goal of the code here is to do the right thing and be as small a footgun as possible until we get this fixed architecturally.  This code is ugly, but it works by expecting the assembler to generate normal code in the mode it's working in and the only instruction encoding fu required by the reader is understanding that 0x90 is a 1-byte NOP.

Reply:
Control register `mov` has a different encoding and doesn't require operand override prefix (66h).

I thought doing that would make things a lot simpler (and this kind of hack isn't all that uncommon either), but it's really up to you :)

Reply:
Sorry, but "This section of assembly gets run in both 32 and 16 bit mode.  Please only use instructions that encode the same in both modes. Kthx." just isn't going to seem "simpler" to anyone who hasn't memorized volume 2 of the IA SDM.  And again, longer term we want the whole locore section to go away.